### ISSUE TITLE
is there a easy way to retrieve blob data to Uint8list

### ISSUE THREAD

**Issue by terryhcw** [hipercompl/fbdb#9]
Hi, 

    I have problem to retrieve two type of blob data from firebird db. The first one is "BLOB sub_type text" and another one is "BLOB sub_type binary".

    For the first one I can not convert _byteBuffer into HTML text or String.

    For the second one which I can retrieve the data and save to file, use the following method, which I think it can have easy way to make it.

      await db2!.startTransaction();
      await q.openCursor(
        sql: "select BLOBDOC "
            "from STORAGEOFDOC where STGUID = ? ",
        parameters: [pid],
        inlineBlobs: false,
      );
      final row = await q.fetchOneAsMap();
      final blobID = row!["BLOBDOC"];
      await db2!.blobToFile(id: blobID, file: File(outFile));
      Uint8List rfile = await File(outFile).readAsBytesSync();
      File(outFile).delete();
      return rfile;

   In this method, which I will create a file and pass the file to other procedure and delete it.

   I had try the stream way, but fail into convert the data type.
   
    Can anyone advise any easy method to work on it?
  
    Regards

Terry Ho

**Comment by ttyrakow** [comment]
You don't need to use `File` as an intermediate representation (apart from the fact that it requires write permission to a file system, it's very inefficient to store a blob into a file just to read it back again as a whole - I understand you did it as a workaround, but there's a better way - keep reading). Instead, you can try this approach (I assume `blobID` is a valid blob ID, obtained as in your code snippet and `db2` is a database attachment and that a transaction has already been started - all just like in your snippet):

```dart
final wholeBlob = List<int>.empty(growable: true);
final blobStream = await db2.openBlob(id: blobID);
await for (final segment in blobStream) {
    // segment is a ByteBuffer, representing a chunk of the blob contents
    wholeBlob.addAll(segment.asUint8List());
}
```

After the `await for` loop completes, `wholeBlob` contains a byte-by-byte representation of your blob contents.
Now you can `utf8.decode` it (if it's a large UTF-8 string) or pass it to any other routine, which requires a list of byte values.

If you need the blob contents to be strictly `Uint8List`, instead of just `List<int>`, you can easily get one from the `List<int>`:
```dart
final newList = Uint8List.fromList(wholeBlob);
```

I don't think you can declare `wholeBlob` itself as `Uint8List`, because (unless I'm mistaken or don't remember correctly) `Uint8List` is not growable (calling `addAll` will throw an exception). That's why you need to keep appending blob segments to a growable `List<int>`, and then convert the complete list to whatever type you need.

Please note, that this approach works as long as the blob fits in the RAM. For *really* large objects you probably need to process data in a segment-by-segment manner (but that's a different story altogether, and not what you asked about).

Is that what you needed? If not, please provide more context, and maybe I'll be able to provide more help.

**Comment by terryhcw** [comment]
Thanks your help. It work, i need more test on different area. I will adopt it the example application.

Also I want to know the db is connected. How can I know it. 

In your example code. I know the  db = await Fbdb.attach(....)
Then it will return db, which print is"Fbdb instance", how can i use it to know the db is connected?

Regards

Terry Ho

**Comment by ttyrakow** [comment]
Well, it depends on what  you mean by "connected". The connection returned by `FbDb.attach` or `FbDb.createDatabase` is always connected (if the process of connecting fails, an exception is thrown and `attach` doesn't return normally at all). However, it can get disconnected (by calling `detach`), so, strictly speaking, a non-null `FbDb` object doesn't necessarily need to represent a connection in an attached state.

Currently there's no dedicated way to test if a network connection to the database is still valid (I'm not even sure Firebird client provides any means to do that, I'll need to check the API), but there are two different approaches, depending on what you actually want to check (what you mean by "connected"): 

1. You can call `await db.ping();` to check if the worker isolate is still running and responding (if not, `ping` will throw an exception), but it's just to test the connection between your main application isolate and the background isolate (native Firebird calls are blocking, so to make a call which would not block your main isolate, a second background isolate is necessary). However, if you call `ping` on a detached connection (i.e. one on which you previously called `await db.detach()`), you'll also get an exception. So `ping` not only just checks if the background isolate is working fine, but also detects `FbDb` instances in a detached state.

2. To check whether the actual network connection to a Firebird server is still on (e.g. wasn't dropped by a network switch, a server crash, a network failure, etc.), you can execute a dummy query of any kind, for example: `await db.selectOne(sql: "select 1 from RDB$DATABASE");`. If it completes without an exception - you're still connected to the database. But if I were you, I wouldn't bother. First, if the network connection is down, you'll get an exception anyway, as soon as you try to execute any kind of SQL statement. Second, the fact that a dummy query worked fine, doesn't guarantee anything. The connection might get dropped right between the dummy query and your real query. The only valid approach is to expect that *every* query may fail at any moment and take appropriate measures (i.e. catch exceptions and handle them).

**Comment by terryhcw** [comment]
Hi,

    Thanks your advise. What I'm working is to make sure when the program is first run, it is success connect to the db, if not, it need the user action to fix it. Like the SQLITE if the db is connect it will be return a connection ID, indicate the program is success connect to the database. What I thinking  is, there will be some value return from the FbDB instance.

     I try to set the Generator value, I use the following statement, which can not success. Is there some different method to make it? 

     try {
        await q.execute(
          sql: "SET GENERATOR GEN_CLIENTADDRESS_ID TO ? ",
          parameters: [100],
        );
      } on FbServerException catch (e) {
        print("error detect : $e");
      }

    Under the above statement, does not return any error in it. 

     While I find it use another method to get the geneator id, just like the folloing statement
     await q.openCursor(
          sql: "select gen_id(gen_clientweb_id,1) from RDB\$DATABASE");
      var r = await q.fetchOneAsMap();
      rval = r!['GEN_ID'];

      the above statement will get the generator ID, while if use the following statement it will not work.
       such as "call gen_clientweb_id(1)

regards

Terry Ho

**Comment by ttyrakow** [comment]
`SET GENERATOR` is part of DDL (data definition language). It cannot be parametrized with `?`. Similarly, you can't execute for example `DROP TABLE ?` and provide the table name as a parameter. Only DQL and DML queries can be dynamically parametrized (it's not a decision made by fbdb, but rather by Firebird itself - you can't prepare a parametrized DDL statement - it always fails). So, in your case, you need to include the generator value in the SQL statement itself:

```dart
await q.execute(sql: "SET GENERATOR GEN_CLIENTADDRESS_ID TO 1");
```

If the value you need to set is in a variable, say `x`, you can always use Dart's standard string interpolation:

```dart
await q.execute(sql: "SET GENERATOR GEN_CLIENTADDRESS_ID TO $x");
```

As for the second part of your question, I don't really understand what you're asking about, please clarify. You say the statement `call gen_clientweb_id(1)` fails, but why shouldn't it? Firebird doesn't understand `call`, it's not a part of Firebird's SQL language (it's a way you call a stored procedure in MySQL for example, but not in Firebird - please refer to the official Firebird SQL manual if you need to determine which statements Firebird understands and which it doesn't).

**Comment by terryhcw** [comment]
Sorry. I checked, it is support by Clarion for window to "call" it.

**Comment by ttyrakow** [comment]
Version 1.4.0 was published at pub.dev today. The Programmer's Guide has been enhanced with an explicit example on how to fetch a blob by segments and put its byte-by-byte representation in an `int` list. Also, there's a whole new section added (4.7) about working with stored procedures, and another (4.8) about handling DML queries with the RETURNING clause.
So I close this issue for now, as there seems nothing more left to explain. 
Feel free to reopen it if you have a follow-up question.
